scipy/optimize/lbfgsb.py

```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
     relationship between the two is ``ftol = factr * numpy.finfo(float).eps``.
     I.e., `factr` multiplies the default machine floating-point precision to
     arrive at `ftol`.
-
+    If any of the lower bounds are equal to the upper bounds then the
+    corresponding entry in ``OptimizeResult.jac`` will be np.nan.
     """
     _check_unknown_options(unknown_options)
     m = maxcor
     pgtol = gtol
     factr = ftol / np.finfo(float).eps
 
-    x0 = asarray(x0).ravel()
-    n, = x0.shape
+    full_x0 = asarray(x0).ravel()
 
     if bounds is None:
-        bounds = [(None, None)] * n
-    if len(bounds) != n:
+        bounds = [(-np.inf, np.inf)] * full_x0.size
```

Comment:
I didn't notice in a first pass why this was needed. Why convert to `inf` here then back to `None`?

Reply:
All the preparation work (`np.clip`, `approx_derivative` (via `ScalarFunction`), `full_new_bounds[0] == full_new_bounds[1]`) requires comparison of numerical arrays. This is why `inf` is used here.

However, LBFGSB Fortran code uses `None` to represent an unbounded lower/upper limit, so any non-finite bounds have to be converted to `None` later on.